### ISSUE TITLE
SMILES augmentation

### ISSUE THREAD

**Issue by sfluegel05** [ChEB-AI/python-chebai#113]
## Idea
There are a lot of different SMILES that describe the same molecule. The SMILES you get depends on different factors: where do you start the traversal? Which direction do you use for rings? How do you sort branches?
So far, we only use the SMILES provided by ChEBI. However, we could get more different SMILES out of ChEBI by applying _SMILES augmentation_, i.e., adding several SMILES to the data set for each molecule. This can potentially reduce overfitting and help generalisation.
## Current state
In PR #52 SMILES augmentation has already been implemented and shown to be successful [see wandb](https://wandb.ai/chebai/chebai/reports/SMILES-augmentation--VmlldzoxMDM5MjAyMw). PR #52 is however outdated.

## Todo
- On a new branch, take the relevant features from #52 and add them to the current chebai
- Use rdkit to generate new SMILES with `rootedAtAtom` and `doRandom`: [rdkit -> MolToSmiles](https://www.rdkit.org/docs/source/rdkit.Chem.rdmolfiles.html)
- The user should be able to specify the number of different SMILES per molecule (as a maximum value, not all molecules might be able to reach that maximum)

**Comment by aditya0by0** [comment]
Setting | Start Atom | Atom Traversal | Output SMILES | Reproducible
-- | -- | -- | -- | --
canonical=True (default) | Auto | Canonical | Canonical | ✅ Yes
doRandom=True | Random | Random | Random | ❌ No (in-session)
rootedAtAtom=X | Atom X | Deterministic | Non-canonical | ✅ Yes
rootedAtAtom=X, doRandom=True | Atom X | Random | Randomized (starting at X) | ❌ No (in-session)

**Comment by aditya0by0** [comment]
The [documentation](https://www.rdkit.org/docs/source/rdkit.Chem.rdmolfiles.html#rdkit.Chem.rdmolfiles.MolToSmiles) says If `doRandom=True`, it **overrides canonical**, meaning it will generate random SMILES no matter what canonical was set to.  
Similarly for `rootedAtAtom` if non-negative, this forces the SMILES to start at a particular atom. Defaults to -1. If not -1, **overrides canonical** setting.

**Comment by sfluegel05** [comment]
That is good to know. In that case, we don't have to worry about canonicalisation here

**Comment by aditya0by0** [comment]
| Setting            | Seed   | Epoch | Train Loss (epoch) | Train Loss (step) | Train Macro-F1 | Train Micro-F1 | Global Step | Val Loss (epoch) | Val Loss (step) | Val Macro-F1 | Val Micro-F1 |
|--------------------|--------|-------|--------------------|-------------------|----------------|----------------|-------------|------------------|-----------------|--------------|--------------|
| Augmentation 15    | 0      | 199   | 0.0007919          | 0.0009062         | 0.9315         | 0.9899         | 1,313,599   | 0.0229558        | 0.0191848       | 0.6850       | 0.9110       |
| Augmentation 15    | 12345  | 199   | 0.0009690          | 0.0003090         | 0.9145         | 0.9881         | 1,313,599   | 0.0210955        | 0.0196686       | 0.6817       | 0.9106       |
| Augmentation 10    | 12345  | 199   | 0.0007078          | 0.0004505         | 0.9390         | 0.9913         | 1,362,799   | 0.0222750        | 0.0389156       | 0.6959       | 0.9129       |
| Augmentation 10    | 0      | 199   | 0.0008145          | 0.0003438         | 0.9294         | 0.9901         | 1,362,799   | 0.0219213        | 0.0530284       | 0.6868       | 0.9123       |
| Augmentation 5     | 12345  | 199   | 0.0009964          | 0.0010712         | 0.9125         | 0.9881         | 747,799     | 0.0185963        | 0.0293011       | 0.6868       | 0.9118       |
| Augmentation 5     | 0      | 199   | 0.0012363          | 0.0013085         | 0.8899         | 0.9853         | 747,799     | 0.0185530        | 0.0249622       | 0.6685       | 0.9088       |
| Original (no aug.) | 0      | 199   | 0.0014627          | 0.0007209         | 0.8564         | 0.9830         | 250,999     | 0.0210648        | 0.0444179       | 0.6031       | 0.8890       |

**Comment by sfluegel05** [comment]
@aditya0by0 Could you add a section to the wiki where you explain (a) how to use augmentation (which flags do I have to set?) and (b) how the augmentation works (how do the SMILES strings look like?), maybe with an example

**Comment by sfluegel05** [comment]
I don't think so. #131 shouldn't affect the results we have here.
